File: ProblemaG/src/G.cpp
#include <bits/stdc++.h>

using namespace std;

struct examen {
	int suma = 0;    // suma de notas de los alumnos en este examen
	int peso;        // peso final del examen
	int peso_max;    // peso máximo admitido
	int ind_entrada; // índice en la entrada

	/**
	 * Comparador de examenes, el menor es el que las notas * en conjunto han sido más bajas. Se usa para ordenar.
	 */
	bool operator<(const examen &ex2) const {
		return suma < ex2.suma;
	};
};

/**
 * Resolver el caso de prueba dado por el vector de exámenes.
 * Imprime el resultado tal como se pide.
 * @param exs vector de exámenes con la suma de notas y las horquillas leídas,
 * @param suma_pesos suma de los pesos asignados en los exámenes.
 * con el peso al mínimo.
 */
void resolver(vector<examen> &exs, int suma_pesos) {
	// ordenamos el vector (de peor resultado a mejor)
	sort(exs.begin(), exs.end());

	// subimos los pesos hasta que suma_pesos = 100
	int index = 0;
	while (suma_pesos < 100) {
		int falta  = 100 - suma_pesos;
		int subida = exs[index].peso_max - exs[index].peso;

		if (subida >= falta) {
			// podemos subir este examen todo lo que queramos
			exs[index].peso += falta;
			break;
		}
		else {
			// subimos lo máximo posible i vamos al próximo examen
			exs[index].peso += subida;
			suma_pesos += subida;
			index++;
		}
	}

	// desordenamos el vector para mostrarlo como nos lo han dado
	vector<int> pesos_orden(exs.size());
	for (examen& ex : exs)
		pesos_orden[ex.ind_entrada] = ex.peso;

	// mostramos el resultado
	for (int i = 0; i < exs.size(); i++) {
		if (i)
			cout << " ";
		cout << pesos_orden[i];
	}
	cout << endl;
}

int main(int argc, char *argv[]) {
	int n, m;
	while (cin >> n >> m) {
		vector<examen> exs(m);

		// leemos las notas de cada alumno
		while (n--) {
			for (int i = 0; i < m; i++) {
				int nota;
				cin >> nota;
				exs[i].suma += nota;
			}
		}

		// leemos las horquillas, asignando la inferior al peso
		int suma_pesos = 0;
		for (int i = 0; i < m; i++) {
			cin >> exs[i].peso >> exs[i].peso_max;
			exs[i].ind_entrada = i;
			suma_pesos += exs[i].peso;
		}

		// resolvemos este caso
		resolver(exs, suma_pesos);
	}
    return 0;
}

File: ProblemaG/src/G.py
def caso(n, m):
    notas = [[0, 0, i] for i in range(m)]   # Notas, límite superior, posición de entrada

    # Almacenar la suma de notas de cada examen
    for a in range(n):  # Exámenes de cada alumno
        e = input().split()

        for i in range(m):
            notas[i][0] += int(e[i])


    # Lista de pesos y suma de los mismos
    p, s = [], 0
    # Almacenar los límites
    for i in range(m):
        inf, sup = [int(i) for i in input().split()]
        notas[i][1] = sup   # Límite superior del examen

        # Empezar utilizando el límite inferior para todos los exámenes
        p.append(inf)
        s += inf
        
    notas = sorted(notas, key = lambda x: x[0]) # Ordenar los exámenes por nota

    # Calcular los pesos
    i = 0
    while s < 100:
        pos = notas[i][2]   # Posición en la lista de pesos del examen actual
        dif = notas[i][1] - p[pos]  # Calcular lo restante hasta el límite superior

        if s+dif <= 100:    # Si se puede sumar todo lo que falta de la horquilla
            s += dif
            p[pos] = notas[i][1]
        else:   # Si con este peso se puede llegar al 100%
            p[pos] = 100-s+p[pos]
            s = 100

        i += 1

    # Mostrar el resultado
    salida = ''
    for elem in p:
        salida += f'{elem} '

    print(salida[:-1])


while True:
    try:
        n, m = [int(i) for i in input().split()]
        caso(n, m)
    except EOFError:
        break
